auslib/util/swagger.py

```diff
@@ -0,0 +1,119 @@
+            elif isinstance(v, dict):
+                self._get_dict(v, key)
+        return {}
+
+    def _load_spec(self, file_name):
+        with file(file_name, 'r') as spec_file:
+            return yaml.load(spec_file)
+
+    def _try_change_version(self, info_dict):
+        version_key = 'version'
+        if info_dict and version_key in info_dict:
+            version = self._get_version()
+            if version:
+                info_dict[version_key] = version
+
+    def add_main_spec(self, file_name):
+        """
+        Adds the current application specification file.
+
+        @param file_name: File relative to app path.
+        @type: string
+        """
+        app_dir = self._app.root_path
+        file_name = path.join(app_dir, file_name)
+        self._main_spec = self._load_spec(file_name)
+        return self
+
+    def add_spec_part(self, file_name):
+        """
+        Adds a part of specification containing either paths, definitions, parameters or both.
+
+        @param file_name: File relative to auslib module to facilitate the reusing of spec files from other places.
+        @type: string
+        """
+        file_name = path.join(AUS_DIR, file_name)
+        spec_part = self._load_spec(file_name)
+
+        if (self.definitions_key not in spec_part and
+                self.endpoints_key not in spec_part and
+                self.parameters_key not in spec_part):
+            raise ValueError("Mergeable swagger keys was not found in: {}".format(file_name))
+
+        self._spec_parts.append(spec_part)
+        return self
+
+    def _test_merge(self, main_dict, part_dict):
+        for k, v in part_dict.items():
+            assert k not in main_dict, "Key {} already exists in spec file.".format(k)
+
+    def _update_endpoints(self, main_endpoints, spec_part):
+        if self.endpoints_key in spec_part:
+            if not main_endpoints:
+                self._main_spec[self.endpoints_key] = main_endpoints
+            part_endpoints = spec_part[self.endpoints_key]
+            self._test_merge(main_endpoints, part_endpoints)
+            main_endpoints.update(part_endpoints)
+
+    def _update_definitions(self, main_definitions, spec_part):
+        if self.definitions_key in spec_part:
+            if not main_definitions:
+                self._main_spec[self.definitions_key] = main_definitions
+            part_definitions = spec_part[self.definitions_key]
+            self._test_merge(main_definitions, part_definitions)
+            main_definitions.update(part_definitions)
+
+    def _update_parameters_definitions(self, main_parameters, spec_part):
+        if self.parameters_key in spec_part:
+            if not main_parameters:
+                self._main_spec[self.parameters_key] = main_parameters
+            part_parameters = spec_part[self.parameters_key]
+            self._test_merge(main_parameters, part_parameters)
+            main_parameters.update(part_parameters)
```

Comment:
Have you considered writing a helper to factor out the common logic of these three methods? Not a blocker or requirement by any means, just a thought.